FlowCrypt/Functionality/Services/EKMService.swift

```diff
@@ -0,0 +1,188 @@
+                    return
+                }
+
+                for keyDetail in keysToUpdate {
+                    passPhrase = try await saveKeyToLocal(
+                        in: viewController,
+                        context: appContext,
+                        keyDetail: keyDetail,
+                        passPhrase: passPhrase
+                    )
+                }
+                await viewController.showToast("refresh_key_success".localized)
+            } catch {
+                if error is ApiError {
+                    return
+                }
+                // Do not display error alert when no keys are returned
+                if let ekmError = error as? EmailKeyManagerApiError, ekmError == .noKeys {
+                    return
+                }
+                // Do not display error alert when user cancels pass phrase prompt
+                if let refreshError = error as? RefreshKeyError, refreshError == .cancelPassPhrase {
+                    return
+                }
+                await viewController.showAlert(message: "refresh_key_error".localizeWithArguments(error.errorMessage))
+            }
+        }
+    }
+
+    private func fetchCurrentPassphrase() throws -> String? {
+        // If this is called when starting the app, then it doesn't make much difference
+        // but conceptually it would be better to look pass phrase both in memory and storage
+        return try appContext.passPhraseService.getPassPhrases(for: appContext.user.email).first(where: { $0.value.isNotEmpty })?.value
+    }
+
+    private func findKeysToUpdate(from keyDetails: [KeyDetails], localKeys: [Keypair]) throws -> [KeyDetails] {
+        var keysToUpdate: [KeyDetails] = []
+        for keyDetail in keyDetails {
+            guard keyDetail.isFullyDecrypted ?? false else { throw EmailKeyManagerApiError.keysAreUnexpectedlyEncrypted }
+            guard let keyLastModified = keyDetail.lastModified else {
+                throw EmailKeyManagerApiError.keysAreInvalid
+            }
+            if let savedLocalKey = localKeys.first(where: { $0.primaryFingerprint == keyDetail.primaryFingerprint }) {
+                if savedLocalKey.lastModified < keyLastModified {
+                    keysToUpdate.append(keyDetail)
+                }
+            } else {
+                keysToUpdate.append(keyDetail)
+            }
+        }
+        return keysToUpdate
+    }
+
+    private func saveKeyToLocal(
+        in viewController: UIViewController,
+        context: AppContextWithUser,
+        keyDetail: KeyDetails,
+        passPhrase: String?
+    ) async throws -> String? {
+        var newPassPhrase = passPhrase
+        if newPassPhrase == nil {
+            if askedUserPassPhrase {
+                return nil
+            }
+            newPassPhrase = try await requestPassPhraseWithModal(in: viewController)
+        }
+        guard let newPassPhrase = newPassPhrase else {
+            return nil
+        }
```

Comment:
This can be removed when we ask for pass phrase in `refreshKeysFromEKMIfNeeded` as mentioned above.